src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.md

```diff
@@ -0,0 +1,389 @@
+densities.
+
+More precisely, the user specifies two sharding parameters:
+
+- base shard count $b$
+- target sstable size $t$
+
+At the start of every compaction, we estimate the density of the output $d$ and calculate a number of shards
+$S$ to split the local token space into to be
+
+$$
+S = 
+\begin{cases}
+2^{\mathrm{round}\left( \log_2 \left( {\frac d t \cdot \frac 1 b}\right)\right)} \cdot b 
+  & \text{if } d \ge tb\\
+b & \text{otherwise}
+\end{cases}
+$$
+
+That is, we divide the density by the target size and round this to a power-of-two multiple of $b$.
+We then generate $S - 1$ boundaries that split the local token space equally into $S$ shards, and split the result
+of the compaction on these boundaries to form a separate sstable for each shard. This aims to produce sstable sizes that 
+fall between $t/\sqrt 2$ and $t\cdot \sqrt 2$.
+
+For example, for a target sstable size of 100MiB and 4 base shards, a 200 MiB memtable will be split in four L0 shards 
+of roughly 50 MiB each, because ${\frac{200}{100} \cdot \frac 1 4} < 1$ and thus we get
+the minimum of 4 shards, each spanning 1/4 of the token space. If in one of these shards we compact 6 of these 50 MiB 
+sstables, the estimated density of the output would be 1200 MiB $({6 \cdot 50 \mathrm{MiB}} / (1/4))$, which results in 
+a target ratio of $\frac{1200}{100} \cdot \frac 1 4 = 2^{\log_2 3}$, rounded to $2^2 \cdot 4$ shards for the whole 
+local token space, thus 4 for the 1/4 span that the compaction covers. Assuming no overwrites and
+deletions, the resulting sstables will be of size 75 MiB, token share 1/16 and density 1200 MiB.
+
+This sharding mechanism is independent of the compaction specification.
+
+## Choosing sstables to compact
+
+The density levelling lets us separate sstables in levels defined by the compaction configuration's fan factors.
+However, unlike in the size levelling case where sstables are expected to cover the full token space, we cannot use the 
+number of sstables on a level as a trigger as many of these sstables may be non-overlapping, i.e. not making read 
+queries less efficient. To deal with this, take advantage of sharding to perform multiple compactions on a level 
+concurrently, and reduce the size of individual compaction operations, we also need to separate non-overlapping 
+sections in different buckets, and decide what to do based on the number of overlapping sstables in a bucket.
+
+To do this, we first form a minimal list of overlap sets that satisfy the following requirements:
+
+- two sstables that do not overlap are never put in the same set;
+- if two sstables overlap, there is a set in the list that contains both;
+- sstables are placed in consecutive positions in the list.
+
+The second condition can also be rephrased to say that for any point in the token range, there is a set in the list
+that contains all sstables whose range covers that point. In other words, the overlap sets give us the maximum number
```

Comment:
As far as I understand, each sstable in a shard is guaranteed to not overlap with any sstables in another shard on the same level, so this set would only contain sstables from the same shard.  If this set is selected for compaction, the result would be multiple sstables on the next level (L+1) that fall in different shards (they do not overlap with each other) but are likely to overlap with existing sstables on L+1.  

Additionally, if there are multiple sets of overlapping sstables that do not contain any of the same sstables, each set can be compacted in parallel.

Is this correct?

Reply:
Yes. Non-overlapping sets can be compacted independently, and we split the results at some predefined positions to guarantee that:
- the sstables in the result don't overlap
- the sstables in the result either fully match the span of any existing sstables in the level where the result is placed, or either the existing or new sstables can be used in pairs (or quads etc) to match the others' span.

A "shard" is not as well defined in UCS-density, because it not only depends on the level, but you can also have different density bands within the same level &mdash; the most common example being levelled compaction, and new sstables entering the span at the lowest density band within the level, but e.g. after two such are combined the result will fall on the next band and will be split in the middle; that's not a problem because we will use both halves when we compact this with the next sstable that enters the level.